### DISCUSSION TITLE
Return string from function

### DISCUSSION THREAD

**Discussion by BigfootN** [fish-shell/fish-shell#10841] [Q&A] [answered]
Hello,

I am desperately trying to return a string from a function. I'm trying to personalize `fish_prompt`

Here is an idea from what I am trying to do:

```fish
function fish_prompt
   if string_length --quiet (my_other_function)
        echo (my_other_function)\n(another_function)
   else
        echo (another_function)
   end
end
```

`my_other_function` looks like this:

```fish
function my_other_function
    string replace <a string> <a string to replace> <the whole string>
end
```

The problem is that I have the result of `my_other_function` printed 3 times (not sure to understand why) for each prompt.

I also tried to add `--quiet` to `string replace` but I face another problem. Nothing is printed.

Thank you very much in advance in advance for any help

**Comment by faho** [comment, marked as answer]
https://fishshell.com/docs/current/language.html#combining-lists-cartesian-product

What you have here is a case where `another_function` prints three lines, and `my_other_function` prints 1.

Because they are directly connected, fish combines each line of another_function with my_other_function.

What you *want* is most likely:

```fish
string join \n -- (my_other_function) (another_function)
```

or just

```fish
my_other_function
another_function
```

without any `echo` at all - these functions already print out output, no need to pass it to something else that just prints it again.

> I also tried to add --quiet to string replace but I face another problem. Nothing is printed.

Okay, if you add `--quiet` that makes your `string replace` not print anything. Which is the opposite of what you want, you're doing `string replace` to get the output in the first place.

----

Sidenote: This construct:

```fish
if string_length --quiet (my_other_function)
        echo (my_other_function)\n(another_function)
```

runs "my_other_function" twice. If you use `string join \n` like above, you can skip the `if` entirely and only run it once.

**Reply by BigfootN** [reply]
Hello,

Thank you for your great answer.

It mostly helped me understand better what was happening.

I still have a question concerning:

```fish
string join \n -- (my_other_function) (another_function)
```

From what I understood from the documentation, `\n` will always be printed ? (in this case `\n` seems to be the separator)

I reread fish's documentation but I didn't find any mentions concerning `--` (in order to better understand), except for `argparse`.

Thank you very much in advance for any help

**Reply by faho** [reply]
> From what I understood from the documentation, \n will always be printed ? (in this case \n seems to be the separator)

\n separates the *arguments* - that means it will be printed in-between each line that `my_other_function` and `another_function` print.

> I reread fish's documentation but I didn't find any mentions concerning -- (in order to better understand), except for argparse.

`--` is a general unix convention. It tells tools that any argument after it isn't an option.

E.g. to search files in the current directory for `--quiet` you would do

```fish
grep -r -- --quiet .
```

If you didn't use `--` here, `grep` would think `--quiet` is an option and interpret "." as the pattern.

---

You're probably also interested in https://fishshell.com/docs/current/prompt.html, which explains the basics of writing a prompt.

**Reply by BigfootN** [reply]
> `--` is a general unix convention. It tells tools that any argument after it isn't an option.

Oh okay it's as simple as that, I thought fish interpreted differently.

Thank you for your detailed answer and help
